File: csharp/interest-is-interesting/InterestIsInteresting.cs
using System;

static class SavingsAccount
{
    public static float InterestRate(decimal balance) =>
        balance switch
        {
            < 0 => 3.213f,
            < 1000 => 0.5f,
            >= 1000 and < 5000 => 1.621f,
            _ => 2.475f
        };

    public static decimal Interest(decimal balance) =>
        (balance / 100) * (decimal)InterestRate(balance);

    public static decimal AnnualBalanceUpdate(decimal balance) =>
        balance + Interest(balance);

    public static int YearsBeforeDesiredBalance(decimal balance, decimal targetBalance)
    {
        var years = 0;
        
        while (balance < targetBalance)
        {
            balance = AnnualBalanceUpdate(balance);
            years++;
        }

        return years;
    }
}

File: cpp/interest-is-interesting/Iteration1/interest_is_interesting.cpp

double interest_rate(double balance) {
    if      (balance >= 5000) { return 2.475; }
    else if (balance >= 1000) { return 1.621; }
    else if (balance >= 0   ) { return 0.5;   }
    else                      { return 3.213; }
}

double yearly_interest(double balance) {
    return (balance / 100) * interest_rate(balance);
}

double annual_balance_update(double balance) {
    return balance + yearly_interest(balance);
}

int years_until_desired_balance(double balance, double target_balance) {
    int years = 0;
    
    while (balance < target_balance) {
        balance = annual_balance_update(balance);
        years++;
    }

    return years;
}
